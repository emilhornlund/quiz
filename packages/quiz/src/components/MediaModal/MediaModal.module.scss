@import '../../styles/colors';
@import '../../styles/helpers';

.mediaModal {
  display: flex;
  flex-direction: column;
  overflow: hidden;

  @include mobile {
    gap: $mobile-control-normal-layout-gutter;
  }

  @include tablet {
    gap: $tablet-control-normal-layout-gutter;
  }

  @include desktop {
    gap: $desktop-control-normal-layout-gutter;
  }

  .column {
    display: flex;
    flex-direction: column;
    justify-content: center;
    width: 100%;

    &.half {
      @include mobile {
        width: calc(50% - $mobile-control-normal-layout-gutter * 0.5);
      }

      @include tablet {
        width: calc(50% - $tablet-control-normal-layout-gutter * 0.5);
      }

      @include desktop {
        width: calc(50% - $desktop-control-normal-layout-gutter * 0.5);
      }
    }

    &.preview {
      display: flex;
      flex-direction: column;
      flex: 1 1 auto;
      min-height: 0;
      overflow: hidden;
    }

    &.inline, &.search {
      flex-direction: row;
      justify-content: space-between;

      @include mobile {
        column-gap: $mobile-control-normal-layout-gutter;
      }

      @include tablet {
        column-gap: $tablet-control-normal-layout-gutter;
      }

      @include desktop {
        column-gap: $desktop-control-normal-layout-gutter;
      }

      .textFieldWrapper {
        flex: 1;
      }
    }

    &.divider {
      border-top-style: dashed;
      border-top-color: $gray-1;

      @include mobile {
        border-top-width: pxToRem(1px);
        margin: $mobile-control-normal-layout-gutter 0;
      }

      @include tablet {
        border-top-width: pxToRem(2px);
        margin: $tablet-control-normal-layout-gutter 0;
      }

      @include desktop {
        border-top-width: pxToRem(3px);
        margin: $desktop-control-normal-layout-gutter 0;
      }
    }

    &.resultsColumn {
      display: flex;
      flex-direction: column;
      flex: 1 1 auto;
      min-height: 0;
      overflow: hidden;

      .resultsGrid {
        display: grid;
        flex: 1 1 auto;
        overflow-y: auto;
        overflow-x: hidden;
        min-height: 0;

        @include mobile {
          gap: $mobile-control-normal-layout-gutter;
          grid-template-columns: 1fr  1fr;
        }

        @include tablet {
          gap: $tablet-control-normal-layout-gutter;
          grid-template-columns: 1fr 1fr  1fr;
        }

        @include desktop {
          gap: $desktop-control-normal-layout-gutter;
          grid-template-columns: 1fr 1fr 1fr 1fr;
        }

        .itemButton {
          display: flex;
          flex-direction: column;
          -moz-appearance: none;
          -o-appearance: none;
          -webkit-appearance: none;
          appearance: none;
          background-color: transparent;
          border: none;
          font-family: inherit;
          outline: none;
          margin: 0;
          padding: 0;
          cursor: pointer;
        }
      }
    }

    &.actions {
      display: flex;
      flex-direction: row;
      flex-wrap: nowrap;
      justify-content: center;
      align-items: center;
      align-self: center;
      width: 100%;
      max-width: $mobile-width;

      @include mobile {
        column-gap: $mobile-control-normal-layout-gutter;
      }

      @include tablet {
        column-gap: $tablet-control-normal-layout-gutter;
      }

      @include desktop {
        column-gap: $desktop-control-normal-layout-gutter;
      }

      div[class*="buttonInputContainer"] {
        flex: 1;
      }
    }
  }
}
