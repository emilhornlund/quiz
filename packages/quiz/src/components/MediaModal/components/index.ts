export type { PexelsImageProviderProps } from './PexelsImageProvider'
export type { UploadImageProviderProps } from './UploadImageProvider'

export { default as PexelsImageProvider } from './PexelsImageProvider'
export { default as UploadImageProvider } from './UploadImageProvider'
