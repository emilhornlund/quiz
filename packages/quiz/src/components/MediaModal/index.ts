export type { MediaModalProps } from './MediaModal'
export { default } from './MediaModal'
