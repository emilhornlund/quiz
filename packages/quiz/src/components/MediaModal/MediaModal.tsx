import { MediaType, QuestionMediaDto, URL_REGEX } from '@quiz/common'
import React, { FC, useMemo, useState } from 'react'

import { MediaTypeLabels } from '../../models/labels.ts'
import { classNames } from '../../utils/helpers.ts'
import Button from '../Button'
import Modal from '../Modal'
import SegmentedControl from '../SegmentedControl'
import Select from '../Select'
import TextField from '../TextField'

import { PexelsImageProvider, UploadImageProvider } from './components'
import styles from './MediaModal.module.scss'

const IMAGE_PROVIDER_PEXELS_VALUE = 'pexels'
const IMAGE_PROVIDER_UPLOAD_VALUE = 'upload'

export interface MediaModalProps {
  title?: string
  imageOnly?: boolean
  type?: MediaType
  url?: string
  onChange: (value?: QuestionMediaDto) => void
  onValid: (valid: boolean) => void
  onClose: () => void
}

const MediaModal: FC<MediaModalProps> = ({
  title,
  imageOnly = false,
  type = MediaType.Image,
  url,
  onChange,
  onValid,
  onClose,
}) => {
  const [internalType, setInternalType] = useState<MediaType>(type)
  const [internalURL, setInternalURL] = useState<string | undefined>(url)

  const [internalValid, setInternalValid] = useState<{
    type: boolean
    url: boolean
  }>({ type: true, url: false })

  const isValid = useMemo(
    () => Object.values(internalValid).every((value) => value),
    [internalValid],
  )

  const handleChangeType = (newType: MediaType) => {
    setInternalType(newType)
    setInternalURL(undefined)
  }

  const onApply = () => {
    if (internalType && internalURL) {
      onChange({ type: internalType, url: internalURL })
      onValid(true)
      onClose()
    }
  }

  const [selectedImageProvider, setSelectedImageProvider] = useState<string>(
    IMAGE_PROVIDER_PEXELS_VALUE,
  )

  return (
    <Modal title={title || 'Add Media'} size="normal" onClose={onClose} open>
      <div className={styles.mediaModal}>
        {!imageOnly && (
          <div className={classNames(styles.column, styles.half)}>
            <Select
              id="media-type-select"
              kind="secondary"
              value={internalType}
              values={Object.values(MediaType).map((type) => ({
                key: type,
                value: type,
                valueLabel: MediaTypeLabels[type],
              }))}
              onChange={(value) => handleChangeType(value as MediaType)}
              onValid={(valid) =>
                setInternalValid({ ...internalValid, type: valid })
              }
              required
            />
          </div>
        )}
        <div className={classNames(styles.column, styles.inline)}>
          <div className={styles.textFieldWrapper}>
            <TextField
              id="media-url-textfield"
              type="text"
              kind="secondary"
              placeholder="URL"
              value={internalURL}
              regex={{ value: URL_REGEX, message: 'Is not a valid URL' }}
              onChange={(value) => setInternalURL(value as string)}
              onValid={(valid) =>
                setInternalValid({ ...internalValid, url: valid })
              }
              required
            />
          </div>
        </div>

        {internalType === MediaType.Image && (
          <>
            <div className={classNames(styles.column, styles.divider)} />
            <div className={styles.column}>
              <SegmentedControl
                id="selected-provider-segmented-control"
                kind="secondary"
                size="small"
                value={selectedImageProvider}
                values={[
                  {
                    key: IMAGE_PROVIDER_PEXELS_VALUE,
                    value: IMAGE_PROVIDER_PEXELS_VALUE,
                    valueLabel: 'Pexels',
                  },
                  {
                    key: IMAGE_PROVIDER_UPLOAD_VALUE,
                    value: IMAGE_PROVIDER_UPLOAD_VALUE,
                    valueLabel: 'Upload',
                  },
                ]}
                onChange={setSelectedImageProvider}
              />
            </div>

            {selectedImageProvider === IMAGE_PROVIDER_PEXELS_VALUE && (
              <PexelsImageProvider onChange={setInternalURL} />
            )}
            {selectedImageProvider === IMAGE_PROVIDER_UPLOAD_VALUE && (
              <UploadImageProvider onChange={setInternalURL} />
            )}
          </>
        )}

        <div className={classNames(styles.column, styles.divider)} />

        <div className={classNames(styles.column, styles.actions)}>
          <Button
            id="close-button"
            type="button"
            kind="secondary"
            value="Close"
            onClick={onClose}
          />
          <Button
            id="apply-button"
            type="button"
            kind="call-to-action"
            value="Apply"
            disabled={!isValid}
            onClick={onApply}
          />
        </div>
      </div>
    </Modal>
  )
}

export default MediaModal
