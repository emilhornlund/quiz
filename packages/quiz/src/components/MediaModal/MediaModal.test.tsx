import { render } from '@testing-library/react'
import React from 'react'
import { describe, expect, it } from 'vitest'

import MediaModal from './MediaModal'

describe('MediaModal', () => {
  it('should render a MediaModal with default props', async () => {
    const { container } = render(
      <MediaModal
        onChange={() => undefined}
        onValid={() => undefined}
        onClose={() => undefined}
      />,
    )

    expect(container).toMatchSnapshot()
  })
})
