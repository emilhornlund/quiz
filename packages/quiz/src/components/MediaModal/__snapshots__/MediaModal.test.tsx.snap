// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MediaModal > should render a MediaModal with default props 1`] = `
<div>
  <div
    class="floatingOverlay"
    style="position: fixed; overflow: auto; top: 0px; right: 0px; bottom: 0px; left: 0px;"
  >
    <span
      data-floating-ui-focus-guard=""
      data-type="inside"
      role="button"
      style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: fixed; white-space: nowrap; width: 1px; top: 0px; left: 0px;"
      tabindex="0"
    />
    <div
      aria-labelledby=":r2:"
      class="modalContainer sizeLarge"
      data-floating-ui-focusable=""
      id=":r0:"
      role="dialog"
      tabindex="-1"
    >
      <div
        class="header"
      >
        <div
          class="title"
          id=":r2:"
        >
          Add Media
        </div>
        <div
          class="buttonInputContainer buttonInputKindPlain"
        >
          <button
            data-testid="test-close-modal-button-button"
            id="close-modal-button"
            name="close-modal-button"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-xmark "
              color="gray"
              data-icon="xmark"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 384 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class="mediaModal"
        >
          <div
            class="column half"
          >
            <div
              class="inputContainer"
            >
              <div
                class="selectInputContainer selectInputKindSecondary"
              >
                <select
                  data-testid="test-media-type-select-select"
                  id="media-type-select"
                  name="media-type-select"
                >
                  <option
                    value="IMAGE"
                  >
                    Image
                  </option>
                  <option
                    value="AUDIO"
                  >
                    Audio
                  </option>
                  <option
                    value="VIDEO"
                  >
                    Video
                  </option>
                </select>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-chevron-down selectIcon"
                  color="black"
                  data-icon="chevron-down"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div
            class="column inline"
          >
            <div
              class="textFieldWrapper"
            >
              <div
                class="inputContainer"
              >
                <div
                  class="textFieldInputContainer textFieldInputKindSecondary"
                >
                  <input
                    class="textfield"
                    data-testid="test-media-url-textfield-textfield"
                    id="media-url-textfield"
                    name="media-url-textfield"
                    placeholder="URL"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="column divider"
          />
          <div
            class="column"
          >
            <div
              class="segmentedControl"
            >
              <div
                class="segmentedControlItem segmentedControlInputKindSecondary segmentedControlInputSizeSmall segmentedControlInputActive"
              >
                <button
                  data-testid="test-selected-provider-segmented-control_pexels-segmented-control"
                  id="selected-provider-segmented-control_pexels"
                  name="selected-provider-segmented-control_pexels"
                  type="button"
                >
                  Pexels
                </button>
              </div>
              <div
                class="segmentedControlItem segmentedControlInputKindSecondary segmentedControlInputSizeSmall segmentedControlInputInactive"
              >
                <button
                  data-testid="test-selected-provider-segmented-control_upload-segmented-control"
                  id="selected-provider-segmented-control_upload"
                  name="selected-provider-segmented-control_upload"
                  type="button"
                >
                  Upload
                </button>
              </div>
            </div>
          </div>
          <div
            class="column divider"
          />
          <div
            class="column actions"
          >
            <div
              class="buttonInputContainer buttonInputKindSecondary"
            >
              <button
                data-testid="test-close-button-button"
                id="close-button"
                name="close-button"
                type="button"
              >
                <span>
                  Close
                </span>
              </button>
            </div>
            <div
              class="buttonInputContainer buttonInputKindCallToAction"
            >
              <button
                data-testid="test-apply-button-button"
                disabled=""
                id="apply-button"
                name="apply-button"
                type="button"
              >
                <span>
                  Apply
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <span
      data-floating-ui-focus-guard=""
      data-type="inside"
      role="button"
      style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: fixed; white-space: nowrap; width: 1px; top: 0px; left: 0px;"
      tabindex="0"
    />
  </div>
</div>
`;
